### PR TITLE
SDK-513 search ads update

### PR DESCRIPTION
**Pull request by echo-branch**
See ticket for background information on this design. 

1.  Used the same delay for first and every retry.  Which is different from spec.
2.  No public API to modify behavior, negotiating with Product on that.
3.  Missing end to end tests around delay, retry and time outs.

### CODE REVIEW

File: Branch-SDK-Tests/BNCAppleSearchAdsTests.m [hunk 1]
```diff
@@ -124,10 +162,38 @@ - (void)testIsTestData_NO {
     XCTAssertFalse([self.appleSearchAds isAppleTestData:testDataIndicators]);
 }
 
+- (void)testIsSearchAdsErrorRetryable_Nil {
+    XCTAssertFalse([self.appleSearchAds isSearchAdsErrorRetryable:nil]);
+}
+
+// ADClientErrorUnknown
+- (void)testIsSearchAdsErrorRetryable_0 {
```
Comment (rkim-branch) [inline Branch-SDK-Tests/BNCAppleSearchAdsTests.m]: I recommend making test name more descriptive.

File: Branch-SDK/Branch-SDK/BNCAppleSearchAds.h [hunk 2]
```diff
@@ -14,10 +14,20 @@ NS_ASSUME_NONNULL_BEGIN
...
 + (BNCAppleSearchAds *)sharedInstance;
 
-// checks Apple Search Ads and updates preferences.  This acquires a lock on BNCPreferenceHelper.
+// Default delay and retry configuration.  ~p90
+// typically less than 1s delay, up to 3.5s delay on first app start
+- (void)useBranchConfig;
+
+// Apple suggests a longer delay, however this is detrimental to app launch times
+// typically less than 1s delay, up to 14s delay on first app start
+- (void)useAppleConfig;
```
Comment (rkim-branch) [inline Branch-SDK/Branch-SDK/BNCAppleSearchAds.h]: Are we locking method name as it is? 

I personally think this method name can be interpreted in multi context, causing confusion to developers.

File: Branch-SDK-Tests/BNCAppleSearchAdsTests.m (hunk 1 above)
Comment (echo-branch) [reply Branch-SDK-Tests/BNCAppleSearchAdsTests.m]: Fixed

File: Branch-SDK/Branch-SDK/BNCAppleSearchAds.m [hunk 3]
```diff
@@ -168,51 +163,116 @@ - (BOOL)isAppleTestData:(NSDictionary *)appleSearchAdDetails {
...
+- (void)requestAttributionWithCompletion:(void (^_Nullable)(NSDictionary *__nullable attributionDetails, NSError *__nullable error, NSTimeInterval elapsedSeconds))completion {
+    
+    // Apple recommends waiting for a short delay between requests for Search Ads Attribution, even the very first request
+    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(self.delay * NSEC_PER_SEC)), dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
 
-    ((void (*)(id, SEL, void (^ __nullable)(NSDictionary *__nullable attrDetails, NSError * __nullable error)))
-    [adClient methodForSelector:self.adClientRequestAttribution])
-    (adClient, self.adClientRequestAttribution, completionBlock);
+        // track timeout
+        __block BOOL raceIsOver = NO;
```
Comment (rkim-branch) [inline Branch-SDK/Branch-SDK/BNCAppleSearchAds.m]: `raceIsOver` -> `completed`

**Comment by rkim-branch** [comment]
👏 Nice job breaking down numerous concurrency logics. It is lot easier to understand, now.

File: Branch-SDK/Branch-SDK/BNCAppleSearchAds.h (hunk 2 above)
Comment (echo-branch) [reply Branch-SDK/Branch-SDK/BNCAppleSearchAds.h]: fixed

File: Branch-SDK/Branch-SDK/BNCAppleSearchAds.m (hunk 3 above)
Comment (echo-branch) [reply Branch-SDK/Branch-SDK/BNCAppleSearchAds.m]: fixed

**Pull request merged**